src/compiler/types.rs

```diff
@@ -64,18 +201,43 @@ impl From<BaseNode> for Node {
 
 #[derive(Debug)]
 pub struct NodeArray {
-    _nodes: Vec<Node>,
+    _nodes: Vec<Rc<Node>>,
 }
 
 impl NodeArray {
-    pub fn new(nodes: Vec<Node>) -> Self {
+    pub fn new(nodes: Vec<Rc<Node>>) -> Self {
         NodeArray { _nodes: nodes }
     }
+
+    pub fn iter(&self) -> NodeArrayIter {
```

Comment:
This stuff I got from [this answer](https://stackoverflow.com/a/40668157/732366) for how to implement an iterator on what is basically a `Vec` wrapper